services/dbaas/dbaas_client.go

```diff
@@ -19,102 +19,476 @@ package dbaas
+}
+
+func (c *Client) watchDbaasControllerEvents() {
+	watchStarted := time.Now()
+	c.subscribeToSupervisordEvents()
+	wg := new(sync.WaitGroup)
+	var cancel context.CancelFunc
+	for {
+		select {
+		case r := <-c.disconnectCh:
+			c.l.Info("Disconnecting from dbaas-controller API.")
+			wg.Wait()
+
+			if c.conn == nil {
+				c.l.Warnf("Trying to disconnect from dbaas-controller API but the connection is not up.")
+				if r.responseCh != nil {
+					r.responseCh <- errorResponse{}
+				}
+				break
+			}
+			err := c.conn.Close()
+			if err != nil {
+				err := errors.Errorf("failed to close conn to dbaas-controller API: %v", err)
+				// Connect request can come from Connect method or from internal watch of supervisord events.
+				// That's why we either return error outside or log it.
+				if r.responseCh != nil {
+					r.responseCh <- errorResponse{err: err}
+					break
+				}
+				c.l.Error(err)
+				break
+			}
+			c.conn = nil
+			if r.responseCh != nil {
+				r.responseCh <- errorResponse{}
+			}
+			c.l.Info("Disconected from dbaas-controller API.")
+		case r := <-c.connectCh:
+			c.l.Infof("Connecting to dbaas-controller API on %s.", c.dbaasControllerAPIAddress)
+			if c.conn != nil {
+				c.l.Warnf("Trying to connect to dbaas-controller API but connection is already up.")
+				if r.responseCh != nil {
+					r.responseCh <- errorResponse{}
+				}
+				break
+			}
+			err := c.connect(r.ctx)
+			// If connect request originates from this watch loop, cancel the request's context.
+			if cancel != nil {
+				cancel()
+				cancel = nil
+			}
+			if err != nil {
+				err := errors.Errorf("failed to connect to dbaas-controller API: %v", err)
+				// Disconnect request can come from Disconnect method or from internal watch of supervisord events.
+				// That's why we either return error outside or log it.
+				if r.responseCh != nil {
+					r.responseCh <- errorResponse{err: err}
+					break
+				}
+				c.l.Error(err)
+			}
+			if r.responseCh != nil {
+				r.responseCh <- errorResponse{}
+			}
+			c.l.Info("Connected to dbaas-controller API.")
+		// Handle supervisord events.
+		case event := <-c.events:
+			c.l.Debugf("Processing supervisord event: %v", event)
+			c.subscribeToSupervisordEvents()
+			if event.Time.Before(watchStarted) {
+				c.l.Warnf("Skipping supervisord event because it was created before the watch of events started: %v", event)
+				break
+			}
+			switch event.Type {
+			case supervisord.Running:
+				var ctx context.Context
+				ctx, localCancel := context.WithTimeout(context.Background(), time.Second*20)
+				select {
+				case c.connectCh <- connectRequest{ctx: ctx}:
+					cancel = localCancel
+				default:
+					localCancel()
+					c.l.Warn("Request to connect to dbaas-controller API was already sent.")
+				}
+			case supervisord.Stopping, supervisord.ExitedUnexpected, supervisord.ExitedExpected:
```

Comment:
I would suggest to disconnect only if `Disconnect()` is called and call it if `DBaaSEnabled` changed to false.
The same about `Connect`. No need to parse supervisord events.
If dbaas-controller fails unexpected, supervisor will restart it, so dbaasClient should retry to connect to dbaas-controller.
If dbaas-controller fails continously then user will see high CPU usage and will be able to disable dbaas by themself.

Reply:
Done. I use rwmutex to protect the conn to the API.